buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -229,18 +229,111 @@ public static int hashCode(ByteBuf buffer) {
      * Returns the reader index of needle in haystack, or -1 if needle is not in haystack.
      */
     public static int indexOf(ByteBuf needle, ByteBuf haystack) {
-        // TODO: maybe use Boyer Moore for efficiency.
-        int attempts = haystack.readableBytes() - needle.readableBytes() + 1;
-        for (int i = 0; i < attempts; i++) {
-            if (equals(needle, needle.readerIndex(),
-                       haystack, haystack.readerIndex() + i,
-                       needle.readableBytes())) {
-                return haystack.readerIndex() + i;
+        if (haystack == null || needle == null) {
+            return -1;
+        }
+
+        if (needle.readableBytes() > haystack.readableBytes()) {
+            return -1;
+        }
+
+        int n = haystack.readableBytes();
+        int m = needle.readableBytes();
+        if (m == 0) {
+            return 0;
+        }
+
+        int i = maxSuf(needle, m, true);
+        int j = maxSuf(needle, m, false);
+        int ell = Math.max(i, j);
+        int memory;
+        int per = 0;
+
+        if (equals(needle, 0, needle, per, per + ell + 1)) {
```

Comment:
I don't understand what this branch is for, because `per` is always zero, so we are just comparing the `needle` to itself for the first `ell + 1` bytes it contains.

Sounds like this would always be true, and indeed the `else` clause has no test coverage.

Reply:
Like the memcmp function of the C language, this judgment is used to compare the first n bytes of the storage area str1 and the storage area str2. If needle does not have a repeated prefix, it will enter the else branch, in fact the possibility of entering the else branch will be bigger

Reply:
This will not check for repeated prefixes, though:

```
if (equals(needle, per, needle, per, per + ell + 1)) {
```

It just checks if the needle is equal to itself in the `per` to `per + per + ell + 1` region, which will either throw an `IndexOutOfBoundsException` or be tautologically true. (Notice that the last argument is a length, not an offset)

Reply:
Like the implementation of http://www-igm.univ-mlv.fr/~lecroq/string/node26.html, use an array of length 2 to return the offset value of the `maxSuf()` method, which does not increase the algorithm Space complexity, but  can use the comparison result of `maxSuf()` method